packages/ckeditor5-ui/src/dropdown/utils.js

```diff
@@ -352,6 +367,39 @@ function focusDropdownContentsOnArrows( dropdownView ) {
 	} );
 }
 
+// Adds a behavior that focuses the #buttonView when the dropdown was closed but focus was within the #panelView element.
+// This makes sure the focus is never lost.
+//
+// @param {module:ui/dropdown/dropdownview~DropdownView} dropdownView
+function focusDropdownButtonOnClose( dropdownView ) {
+	dropdownView.on( 'change:isOpen', ( evt, name, isOpen ) => {
+		if ( isOpen ) {
+			return;
+		}
+
+		// If the dropdown was closed, move the focus back to the button (#12125).
+		// Don't touch the focus, if it moved somewhere else (e.g. moved to the editing root on #execute) (#12178).
+		// Note: Don't use the state of the DropdownView#focusTracker here. It fires #blur with the timeout.
+		if ( dropdownView.panelView.element.contains( global.document.activeElement ) ) {
+			dropdownView.buttonView.focus();
+		}
+	} );
+}
+
+// Adds a behavior that focuses the #panelView when dropdown gets open (accessibility).
+//
+// @param {module:ui/dropdown/dropdownview~DropdownView} dropdownView
+function focusDropdownPanelOnOpen( dropdownView ) {
+	dropdownView.on( 'change:isOpen', ( evt, name, isOpen ) => {
+		if ( !isOpen ) {
+			return;
+		}
+
+		// Focus the first item in the dropdown when the dropdown opened.
+		dropdownView.panelView.focus();
+	}, { priority: 'low' } ); // Let the panel show up first.
```

Comment:
This listener works on the same event and priority as in `focusChildOnDropdownOpen()` function. I checked and the order of calls is correct (at first the panel is focused and then the focus is moved to the child), but at the first glance it looks like a race condition or coincidence. I'd rather set the priority to some arbitrary number above the `low` priority like `-100` or at least add a comment that "it will always be executed before `focusChildOnDropdownOpen()` so don't worry child".

Reply:
Nice catch 🙂 Still, the order of execution will always be safe because `focusDropdownPanelOnOpen()` is always called first (when the dropdown is created) and `focusChildOnDropdownOpen()` is called next by the integrator and if the priorities are the same, callbacks are executed in the same order they were registered in the `EmitterMixin`.

I suppose `focusDropdownPanelOnOpen, focusChildOnDropdownOpen` this is the right order of things, even if they both get called. 

If it was the other way around (`focusChildOnDropdownOpen, focusDropdownPanelOnOpen`), the focus (in the headings dropdown) would first move to the selected heading, then back to the entire panel, which would focus the first item in the list, which is not necessarily the active heading.

Reply:
I'll change the priorities just to keep things easier to understand without the knowledge about internals of `EmitterMixin`.

Reply:
Addressed in d096b11.